File: src/main/java/io/codeka/gaia/stacks/service/RegexStackVariablesValidator.java
package io.codeka.gaia.stacks.service;

import io.codeka.gaia.modules.repository.TerraformModuleRepository;
import io.codeka.gaia.stacks.bo.RegexStackVariablesValidation;
import io.codeka.gaia.stacks.bo.Stack;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Service;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import java.util.regex.Pattern;

@Service
public class RegexStackVariablesValidator implements ConstraintValidator<RegexStackVariablesValidation, Stack> {

    @Autowired
    private TerraformModuleRepository moduleRepository;

    @Override
    public void initialize(RegexStackVariablesValidation constraintAnnotation) {
        // nothing to do here
    }

    @Override
    public boolean isValid(Stack stack, ConstraintValidatorContext constraintValidatorContext) {
        if(stack.getModuleId() == null || stack.getModuleId().isBlank()){
            // no module defined, so variables are valid !
            return true;
        }
        // getting the module
        var module = this.moduleRepository.findById(stack.getModuleId()).orElseThrow();

        return module.getVariables().stream()
                // not checking non regex values
                .filter(terraformVariable -> terraformVariable.getValidationRegex() != null)
                // not checking blank values (doing it first because Pair want non-null values)
                .filter(terraformVariable -> {
                    var value = stack.getVariableValues().get(terraformVariable.getName());
                    return value != null && ! value.isBlank();
                })
                // associating each variable with its value
                .map(terraformVariable -> Pair.of(terraformVariable, stack.getVariableValues().get(terraformVariable.getName())))
                .allMatch(pair -> {
                    var pattern = Pattern.compile(pair.getFirst().getValidationRegex());
                    return pattern.matcher(pair.getSecond()).matches();
                });
    }
}

File: src/main/java/io/codeka/gaia/stacks/service/RegexStackVariablesValidator.kt
package io.codeka.gaia.stacks.service

import io.codeka.gaia.modules.bo.Variable
import io.codeka.gaia.modules.repository.TerraformModuleRepository
import io.codeka.gaia.stacks.bo.RegexStackVariablesValidation
import io.codeka.gaia.stacks.bo.Stack
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import java.util.regex.Pattern
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

@Service
class RegexStackVariablesValidator(@Autowired val moduleRepository: TerraformModuleRepository)
    : ConstraintValidator<RegexStackVariablesValidation, Stack> {

    override fun isValid(stack: Stack, constraintValidatorContext: ConstraintValidatorContext?): Boolean {
        // we do not validate anything if the stacks has no module reference (it will send a BadRequest anyway)
        if( stack.moduleId.isNullOrBlank()) {
            return true
        }
        // getting the module, returning true if module cannot be found !
        // this should (let's hope so) never happen ! another validator should handle this case
        val module = this.moduleRepository.findByIdOrNull(stack.moduleId) ?: return true

        return module.variables
                // not checking non regex values
                .filter { ! it.validationRegex.isNullOrBlank() }
                // not checking blank values (doing it first because Pair want non-null values)
                .filter { ! stack.variableValues[it.name].isNullOrBlank() }
                // validate the pattern (using non null assert as variable value is not more null here)
                .all { it.validateWithPattern(stack.variableValues[it.name]!!) }
    }
}

private fun Variable.validateWithPattern(value: String): Boolean {
    val pattern = Pattern.compile(this.validationRegex)
    return pattern.matcher(value).matches()
}
